keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,585 @@ yet but we think may be interesting to consider in the future.
+```
+
+### Default Behavior
+
+There must be reasonable behavior "out of the box", and it should be
+at least a little difficult for an administrator to lock himself out
+of this subsystem.  To accomplish these things there are two levels of
+defaulting: one concerning behavior, and one concerning explicit API
+objects.
+
+The effective configuration is the union of (a) the actual API objects
+that exist and (b) implicitly generated backstop objects.  The latter
+are not actual API objects, and might not ever exist as identifiable
+objects in the implementation, but are figments of our imagination
+used to describe the behavior of this subsystem.  These backstop
+objects are implicitly present and affecting behavior when needed.
+There are two implicitly generated RequestPriority backstop objects.
+One is equivalent to the `system-top` object exhibited above, and it
+exists while there is no actual RequestPriority object with `exempt ==
+true`.  The other is equivalent to the `workload-low` object exhibited
+above, and exists while there is no RequestPriority object with
+non-exempt priority.  There are also two implicitly generated
+FlowSchema backup objects.  Whenever a request whose groups include
+`system:masters` is not matched by any actual FlowSchema object, a
+backstop equivalent to the `system-top` object exhibited above is
+considered to exist.  Whenever a request whose groups do not include
+`system:masters` is not matched by any actual FlowSchema object, the
+following backstop object is considered to exist.
+
+```yaml
+kind: FlowSchema
+meta:
+  name: non-top-backstop
+spec:
+  matchingPriority: (doesn’t really matter)
+  requestPriority:
+    name: (name of an effectively existing RequestPriority, whether
+	that is real or backstop, with catchAll==true)
+  flowDistinguisher:
+    source: user
+    # no transformation in this case
+  match:
+  - and: [ ] # match everything
+```
+
+The other part of the defaulting story concerns making actual API
+objects exist, and it goes as follows.  Whenever there is no actual
+RequestPriority object with `exempt == true`, the RequestPriority
+objects exhibited above are created --- except those with a name
+already in use by an existing RequestPriority object.  Whenever there
+is no actual FlowSchema object that refers to an exempt
+RequestPriority object, the schema objects shown above as examples are
+generated --- except those with a name already in use.
+
+### Prometheus Metrics
+
+Prior to this KEP, the relevant available metrics from an apiserver are:
+- apiserver_current_inflight_requests (gauge, broken down by mutating or not)
+- apiserver_longrunning_gauge
+- apiserver_request_count (cumulative number served)
+- apiserver_request_latencies (histogram)
+- apiserver_request_latencies_summary
+
+This KEP adds the following metrics.
+- apiserver_rejected_requests (count, broken down by priority, FlowSchema, when (arrival vs timeout))
```

Comment:
Did we discuss whether including flowschema is too high-cardinality? I think it should be OK if we don't multiply it by too many other dimensions. It might be good to state the expected cardinality here in a followup; ending up with too many distinct label values is a common failure mode in metrics.

Reply:
Note that what is stipulated here is the schema, not the individual flow.  There is a statement elsewhere about the expected cardinality of schemas.